packages/yoastseo/src/languageProcessing/researches/countAnchorsWithKeyphrase.js

```diff
@@ -0,0 +1,217 @@
+ * Also includes different word forms if the morphology is available.
+ *
+ * @param {Paper}       paper       The paper to research.
+ * @param {Researcher}  researcher  The researcher to use.
+ *
+ * @returns {Object} The amount of anchor texts whose content words are the same as the keyphrase or synonyms' content words.
+ */
+export default function( paper, researcher ) {
+	functionWords = researcher.getConfig( "functionWords" );
+	const result = {
+		anchorsWithKeyphrase: [],
+		anchorsWithKeyphraseCount: 0,
+	};
+	// STEP 1.
+	// If the paper's text is empty, return empty result.
+	if ( paper.getText() === "" ) {
+		return result;
+	}
+
+	// STEP 2.
+	const keyphrase = paper.getKeyword();
+	/*
+	 * If no keyphrase is set, return empty result.
+	 * This is a conscious decision where we won't assess the paper if the keyphrase is not set.
+	 * This includes a case where only the synonym is set but not the keyphrase.
+	 */
+	if ( keyphrase === "" ) {
+		return result;
+	}
+	/*
+     * When the keyphrase is set, also retrieve the synonyms and save them in "topics" array.
+     * Eventually, the term topics here refers to either keyphrase or synonyms.
+    */
+	const originalTopics = parseSynonyms( paper.getSynonyms() );
+	originalTopics.push( keyphrase );
+
+	// Retrieve the anchors.
+	let anchors = paper.getTree().findAll( treeNode => treeNode.name === "a" );
+	/*
+	 * We get the site's URL (e.g., https://yoast.com) or domain (e.g., yoast.com) from the paper.
+	 * In case of WordPress, the variable is a URL. In case of Shopify, it is a domain.
+	 */
+	const siteUrlOrDomain = paper.getPermalink();
+
+	// STEP 3.
+	// Filter anchors with urls that are not linking to the current site url/domain.
+	anchors = filterAnchorsLinkingToSelf( anchors, siteUrlOrDomain );
+	// If all anchor urls are linking to the current site url/domain, return empty result.
+	if ( anchors.length === 0 ) {
+		return result;
+	}
+
+	const locale = paper.getLocale();
+	const topicForms = researcher.getResearch( "morphology" );
+	const customHelpers = {
+		matchWordCustomHelper: researcher.getHelper( "matchWordCustomHelper" ),
+		getWordsCustomHelper: researcher.getHelper( "getWordsCustomHelper" ),
+	};
+
+	// STEP 4.
+	// Filter anchors with text that contains the keyphrase/synonyms' content words.
+	anchors = filterAnchorsContainingTopic( anchors, topicForms, locale, customHelpers.matchWordCustomHelper );
+	// If all anchor texts do not contain the keyphrase/synonyms' content words, return empty result.
+	if ( anchors.length === 0 ) {
+		return result;
+	}
+
+	// STEP 5.
+	// Check if exact match is requested for every topic (keyphrase or synonym).
+	const isExactMatchRequested = originalTopics.map( originalTopic => processExactMatchRequest( originalTopic ) );
+	// Filter anchors with text that has the same content words as the keyphrase/synonyms.
+	anchors = filterAnchorsContainedInTopic( anchors, topicForms, locale, customHelpers, isExactMatchRequested );
```

Comment:
If the condition in each step (1-4) is fulfilled, the empty result is returned. This way, we don't need to execute the next line of code (if the condition is not fulfilled).

Variables are also declared only when they are needed in each step. This way, we don't declare a variable when it's eventually not being used.